src/Microsoft.NetCore.Analyzers/Core/Security/DoNotOverloadDeserializationBinderWithoutThrowingAnException.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.NetCore.Analyzers.Security.Helpers;
+
+namespace Microsoft.NetCore.Analyzers.Security
+{
+    /// <summary>
+    /// Checks overloads of the serialization binder's BindToType method
+    /// to check whether they properly throw an exception
+    /// </summary>
+    /// <remarks>
+    /// 
+    /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class DoNotOverloadSerializationBinderWithoutThrowingAnException : DiagnosticAnalyzer
+    {
+        protected string OverloadingClassMetadataName { get; }
+        enum CacheState
+        {
+            Absent,
+            InProgress,
+            PresentTrue,
+            PresentFalse
+        };
+
+        internal static DiagnosticDescriptor DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule = SecurityHelpers.CreateDiagnosticDescriptor(
+            "CA2340",
+            nameof(MicrosoftNetCoreAnalyzersResources.DoNotOverloadSerializationBinderWithoutThrowingAnExceptionTitle),
+            nameof(MicrosoftNetCoreAnalyzersResources.DoNotOverloadSerializationBinderWithoutThrowingAnExceptionMessage),
+            DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create<DiagnosticDescriptor>(
+                DoNotOverloadSerializationBinderWithoutThrowingAnExceptionRule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    Dictionary<IMethodSymbol, CacheState> methodCache = new Dictionary<IMethodSymbol, CacheState>();
```

Comment:
>Dictionary [](start = 20, length = 10)

@mavasani I think we need to worry about thread safety, right?

Could use a ConcurrentDictionary instead.

Reply:
Yes, certainly needs a Concurrent Dictionary